### PR TITLE
Make the CheckpointVerifier handle partial restarts

### LINKED ISSUE #732: Make the CheckpointVerifier handle partial restarts
The CheckpointVerifier assumes that we always start at the genesis block.

But that's not true if we do a partial chain sync, and then get interrupted during a checkpoint.

Dependencies:
- [ ] rebase PR on the latest #731
- [ ] get PR #731 merged
- [ ] rebase this PR on main

Testing:
- [ ] test mainnet syncing

### PR DESCRIPTION
**Pull request by teor2345**
Closes #732.

This PR is based on #731, but feel free to merge them both if they're ready.

The first new commit is 5696b4e.

### CODE REVIEW

File: zebra-consensus/src/checkpoint.rs
Comment (yaahc) [inline zebra-consensus/src/checkpoint.rs:326]: mmmm, I love the smell of or patterns in the morning.

File: zebra-consensus/src/checkpoint/tests.rs
Comment (yaahc) [inline zebra-consensus/src/checkpoint/tests.rs:298]: Wondering if this could be expressed as a `zebra_test::Transcript`

It seems like the only thing that's missing is some sort of customizable callback to do additional assertions after each request. I think we might be able to bring in the assertions as an additional service wrapping the transcript.

File: zebra-consensus/src/checkpoint/types.rs
Comment (yaahc) [inline zebra-consensus/src/checkpoint/types.rs:49]: wow, that's definitely a pattern!

I wonder if it would also be valid to write it like this

`(InitialTip(self_height) | PreviousCheckpoint(self_height), InitialTip(other_height) | PreviousCheckpoint(other_height)) => ...`

Edit: GAH, it's almost valid but still experimental

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=249eede9b3e9324b6a0d22abda7bb3de

https://github.com/rust-lang/rust/issues/54883

some day we will have all of the nice things :/

File: zebrad/src/commands/start.rs
Comment (yaahc) [inline zebrad/src/commands/start.rs]: I might be missing it but this doesn't seem to be accessing any state of the `StartCmd`. I think it might make more sense to make this a method on the `Verifier` and have it grab the tip from the provided state as part of its constructor, rather than passing it in as an arg.

File: zebra-consensus/src/checkpoint/tests.rs
Comment (teor2345) [reply zebra-consensus/src/checkpoint/tests.rs:298]: Would you like to open another ticket?

I'm not sure where to start here.

File: zebrad/src/commands/start.rs
Comment (teor2345) [reply zebrad/src/commands/start.rs]: How can I execute async functions in a non-async constructor?

File: zebra-consensus/src/checkpoint/tests.rs
Comment (yaahc) [reply zebra-consensus/src/checkpoint/tests.rs:298]: Yea, I'll open another issue, no point holding this up.

File: zebrad/src/commands/start.rs
Comment (yaahc) [reply zebrad/src/commands/start.rs]: oh yea, I'd see if you can make the constructor async

Comment (teor2345) [reply zebrad/src/commands/start.rs]: There should be a clippy lint for async, template functions that don't actually use self.

Comment (teor2345) [reply zebrad/src/commands/start.rs]: That function doesn't actually depend on any verifier state either, so I moved it to zebra-state in 6c987e036e88ec34bb7c03f2275eae1f2fa5ee9a.

I imagine you'll want to rewrite it as part of the chain reorg changes.

**Review by yaahc — CHANGES_REQUESTED** [review]
Looks good to go once the conflicts are fixed.

**Pull request merged**
